File: worker/src/scorecard.js
// scorecard.js — golf hole-by-hole player summary → canonical
// GolfScorecardResponse (see canonical.ts). Pure (no I/O); the raw payload is
// the web-host `leaderboard/{event}/playersummary` (VERIFIED 2026-07):
//   profile — id/displayName/headshot(url string)/…
//   rounds[] — per round: value(strokes)/displayValue(to-par)/inScore/outScore/
//              teeTime/startTee/groupNumber/currentPosition + linescores[] per
//              hole {period, value, par, scoreType{name}}. Future rounds ship
//              teeTime with no holes — that IS the pre-start glance.
//   stats[] — tournament stat line; we keep a small curated subset.

import { https } from './normalize.js';

const pick = (o, keys) => Object.fromEntries(keys.filter(k => o[k] != null && o[k] !== '').map(k => [k, o[k]]));

// The glanceable tournament stats — everything else (strokes gained tables etc.)
// is analyst depth the product excludes.
const STAT_KEEP = new Set([
  'scoreToPar', 'regScore', 'birdies', 'eagles',
  'driveDistAvg', 'driveAccuracyPct', 'greensInRegulationPct', 'puttsPerHole',
]);

export function normalizeGolfScorecard(key, eventId, playerId, raw) {
  const prof = raw?.profile || {};
  const player = pick({
    id: String(prof.id ?? playerId),
    name: prof.displayName || prof.shortName || '',
    headshot: https(prof.headshot?.href || prof.headshot),
    country: prof.flag?.alt || prof.citizenship,
  }, ['id', 'name', 'headshot', 'country']);

  const rounds = (Array.isArray(raw?.rounds) ? raw.rounds : []).map(r => {
    const holes = (Array.isArray(r.linescores) ? r.linescores : []).map(h => pick({
      hole: typeof h.period === 'number' ? h.period : undefined,
      par: typeof h.par === 'number' ? h.par : undefined,
      strokes: typeof h.value === 'number' ? h.value : undefined,
      scoreType: h.scoreType?.name, // 'BIRDIE' | 'PAR' | 'BOGEY' …
    }, ['hole', 'par', 'strokes', 'scoreType'])).filter(h => h.hole != null);
    const round = pick({
      round: typeof r.period === 'number' ? r.period : undefined,
      strokes: typeof r.value === 'number' && r.value > 0 ? r.value : undefined,
      toPar: r.displayValue != null && r.displayValue !== '-' ? String(r.displayValue) : undefined,
      outScore: typeof r.outScore === 'number' && r.outScore > 0 ? r.outScore : undefined,
      inScore: typeof r.inScore === 'number' && r.inScore > 0 ? r.inScore : undefined,
      teeTime: r.teeTime,
      startTee: typeof r.startTee === 'number' ? r.startTee : undefined,
      groupNumber: typeof r.groupNumber === 'number' ? r.groupNumber : undefined,
      currentPosition: typeof r.currentPosition === 'number' ? r.currentPosition : undefined,
    }, ['round', 'strokes', 'toPar', 'outScore', 'inScore', 'teeTime', 'startTee', 'groupNumber', 'currentPosition']);
    round.holes = holes; // [] pre-round, per contract
    return round;
  }).filter(r => r.round != null);

  const stats = (Array.isArray(raw?.stats) ? raw.stats : [])
    .filter(s => STAT_KEEP.has(s.name) && s.displayValue != null && s.displayValue !== '')
    .map(s => ({ name: s.name, label: s.displayName || s.name, value: String(s.displayValue) }));

  const out = { league: key, eventId: String(eventId), player, rounds };
  if (stats.length) out.stats = stats;
  return out;
}

File: app/lib/src/data/scorecard.dart
// scorecard.dart — Dart port of worker/src/scorecard.js. Golf hole-by-hole
// player summary → canonical GolfScorecardResponse. Pure.

import 'util.dart';

const _statKeep = {
  'scoreToPar', 'regScore', 'birdies', 'eagles',
  'driveDistAvg', 'driveAccuracyPct', 'greensInRegulationPct', 'puttsPerHole',
};

Map<String, dynamic> normalizeGolfScorecard(String key, dynamic eventId, dynamic playerId, dynamic raw) {
  final prof = field(raw, 'profile') ?? {};
  final player = pickT({
    'id': jsStr(field(prof, 'id') ?? playerId),
    'name': or([field(prof, 'displayName'), field(prof, 'shortName'), '']),
    'headshot': https(or([field(field(prof, 'headshot'), 'href'), field(prof, 'headshot')])),
    'country': or([field(field(prof, 'flag'), 'alt'), field(prof, 'citizenship')]),
  }, ['id', 'name', 'headshot', 'country']);

  final rounds = <Map<String, dynamic>>[];
  for (final r in (field(raw, 'rounds') is List ? field(raw, 'rounds') as List : const [])) {
    final holes = <Map<String, dynamic>>[];
    for (final h in (field(r, 'linescores') is List ? field(r, 'linescores') as List : const [])) {
      final hole = pickT({
        'hole': field(h, 'period') is num ? field(h, 'period') : null,
        'par': field(h, 'par') is num ? field(h, 'par') : null,
        'strokes': field(h, 'value') is num ? field(h, 'value') : null,
        'scoreType': field(field(h, 'scoreType'), 'name'),
      }, ['hole', 'par', 'strokes', 'scoreType']);
      if (hole['hole'] != null) holes.add(hole);
    }
    final round = pickT({
      'round': field(r, 'period') is num ? field(r, 'period') : null,
      'strokes': (field(r, 'value') is num && (field(r, 'value') as num) > 0) ? field(r, 'value') : null,
      'toPar': (field(r, 'displayValue') != null && field(r, 'displayValue') != '-') ? jsStr(field(r, 'displayValue')) : null,
      'outScore': (field(r, 'outScore') is num && (field(r, 'outScore') as num) > 0) ? field(r, 'outScore') : null,
      'inScore': (field(r, 'inScore') is num && (field(r, 'inScore') as num) > 0) ? field(r, 'inScore') : null,
      'teeTime': field(r, 'teeTime'),
      'startTee': field(r, 'startTee') is num ? field(r, 'startTee') : null,
      'groupNumber': field(r, 'groupNumber') is num ? field(r, 'groupNumber') : null,
      'currentPosition': field(r, 'currentPosition') is num ? field(r, 'currentPosition') : null,
    }, ['round', 'strokes', 'toPar', 'outScore', 'inScore', 'teeTime', 'startTee', 'groupNumber', 'currentPosition']);
    round['holes'] = holes;
    if (round['round'] != null) rounds.add(round);
  }

  final stats = (field(raw, 'stats') is List ? field(raw, 'stats') as List : const [])
      .where((s) => _statKeep.contains(field(s, 'name')) && field(s, 'displayValue') != null && field(s, 'displayValue') != '')
      .map((s) => {'name': field(s, 'name'), 'label': or([field(s, 'displayName'), field(s, 'name')]), 'value': jsStr(field(s, 'displayValue'))})
      .toList();

  final out = <String, dynamic>{'league': key, 'eventId': jsStr(eventId), 'player': player, 'rounds': rounds};
  if (stats.isNotEmpty) out['stats'] = stats;
  return out;
}
